File: step_08_bit_manipulation/8.2_interview_problems/P002_power_set_using_bits/solution.py
"""
Problem: Power Set using Bits
Difficulty: MEDIUM | XP: 25
"""
from typing import List


# ============================================================
# APPROACH 1: BRUTE FORCE — Recursive generation
# Time: O(2^n * n)  |  Space: O(2^n * n)
# ============================================================
def _recurse(nums: List[int], index: int,
             current: List[int], result: List[List[int]]) -> None:
    if index == len(nums):
        result.append(current[:])
        return
    # Include nums[index]
    current.append(nums[index])
    _recurse(nums, index + 1, current, result)
    current.pop()
    # Exclude nums[index]
    _recurse(nums, index + 1, current, result)


def brute_force(nums: List[int]) -> List[List[int]]:
    """
    Recursively decide include/exclude for each element.
    Generates all 2^n subsets.
    """
    result: List[List[int]] = []
    _recurse(nums, 0, [], result)
    return result


# ============================================================
# APPROACH 2: OPTIMAL — Bitmask iteration
# Time: O(n * 2^n)  |  Space: O(n * 2^n) for storing subsets
# For each number 0 .. 2^n-1, treat its bits as an inclusion
# mask: bit j set → include nums[j].
# ============================================================
def optimal(nums: List[int]) -> List[List[int]]:
    """
    Iterate over all 2^n masks.  For each mask, collect elements
    whose corresponding bit is set.
    """
    n = len(nums)
    result: List[List[int]] = []
    for mask in range(1 << n):          # 0 to 2^n - 1
        subset = []
        for j in range(n):
            if mask & (1 << j):
                subset.append(nums[j])
        result.append(subset)
    return result


# ============================================================
# APPROACH 3: BEST — Bitmask, bit-count ordered (lexicographic)
# Time: O(n * 2^n)  |  Space: O(n * 2^n)
# Same complexity, but groups subsets by size using Python's
# built-in bin().count('1') — natural for interview readability.
# ============================================================
def best(nums: List[int]) -> List[List[int]]:
    """
    Bitmask approach, subsets ordered by cardinality then value
    for a clean presentation.
    """
    n = len(nums)
    all_masks = sorted(range(1 << n), key=lambda m: bin(m).count('1'))
    result: List[List[int]] = []
    for mask in all_masks:
        subset = [nums[j] for j in range(n) if mask & (1 << j)]
        result.append(subset)
    return result


if __name__ == "__main__":
    print("=== Power Set using Bits ===")
    nums = [1, 2, 3]
    print(f"Input: {nums}")
    print(f"Brute ({len(brute_force(nums))} subsets):   {brute_force(nums)}")
    print(f"Optimal ({len(optimal(nums))} subsets): {optimal(nums)}")
    print(f"Best ({len(best(nums))} subsets, size-ordered): {best(nums)}")

File: step_08_bit_manipulation/8.2_interview_problems/P002_power_set_using_bits/Solution.java
/**
 * Problem: Power Set using Bits
 * Difficulty: MEDIUM | XP: 25
 *
 * @author DSA_Nova
 */
import java.util.*;

public class Solution {

    // ============================================================
    // APPROACH 1: BRUTE FORCE - Recursive pick/not-pick
    // Time: O(2^n * n)  |  Space: O(n) recursion depth
    // ============================================================
    public static List<List<Integer>> bruteForce(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        generateSubsets(nums, 0, new ArrayList<>(), result);
        return result;
    }

    private static void generateSubsets(int[] nums, int idx, List<Integer> current,
                                        List<List<Integer>> result) {
        if (idx == nums.length) {
            result.add(new ArrayList<>(current));
            return;
        }
        // Don't pick
        generateSubsets(nums, idx + 1, current, result);
        // Pick
        current.add(nums[idx]);
        generateSubsets(nums, idx + 1, current, result);
        current.remove(current.size() - 1);
    }

    // ============================================================
    // APPROACH 2: OPTIMAL - Bitmask iteration
    // Time: O(2^n * n)  |  Space: O(1) extra (excluding output)
    // ============================================================
    public static List<List<Integer>> optimal(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        int n = nums.length;
        int total = 1 << n; // 2^n
        for (int mask = 0; mask < total; mask++) {
            List<Integer> subset = new ArrayList<>();
            for (int i = 0; i < n; i++) {
                if ((mask & (1 << i)) != 0) {
                    subset.add(nums[i]);
                }
            }
            result.add(subset);
        }
        return result;
    }

    // ============================================================
    // APPROACH 3: BEST - Iterative build (cascading)
    // Time: O(2^n * n)  |  Space: O(1) extra (excluding output)
    // ============================================================
    public static List<List<Integer>> best(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        result.add(new ArrayList<>()); // start with empty set
        for (int num : nums) {
            int size = result.size();
            for (int i = 0; i < size; i++) {
                List<Integer> newSubset = new ArrayList<>(result.get(i));
                newSubset.add(num);
                result.add(newSubset);
            }
        }
        return result;
    }

    public static void main(String[] args) {
        System.out.println("=== Power Set using Bits ===");
        int[] nums = {1, 2, 3};
        System.out.println("Brute Force: " + bruteForce(nums));
        System.out.println("Optimal:     " + optimal(nums));
        System.out.println("Best:        " + best(nums));
    }
}
